"Monthly" Daily scheduling added to API

This is stemming from an issue in the mobile apps where the scheduling of a monthly Daily won't update correctly if the start date is changed. It was decided that it'd be better to fix it through the API than locally. 

You can see the detailed report from the original issue here: https://github.com/HabitRPG/habitica-android/issues/1272

I would appreciate some help debugging an issue that I've encountered in my local development environment if anyone has a moment to spare (and has experienced similar issues in a Windows 10 Docker local development setup).

I posted a question on the [Aspiring Blacksmiths guild](https://habitica.com/groups/guild/68d4a01e-db97-4786-8ee3-05d612c5af6f)... I'll copy the message here as well since I'm not sure if I can permalink the comment.

### Local Development git repository issue that I could use help with (copied from Aspiring Blacksmiths guild)

Hello friends,

I have been working on some Habitica issues. I'm trying to get really ramped-up into the Habitica API; API development is really enjoyable to me 😁👍. (My account is [@terminalstderr](https://github.com/terminalstderr) on github and I'm currently working on ["Monthly" Daily scheduling added to API #12041](https://github.com/HabitRPG/habitica/issues/12041))

I've run into an issue with my local development environment. Everything was working well at first, but now I'm stumbling into a really strange behavior with a git repository that needs to be accessed by my Client Docker container.

The issue is that my Client Docker container kicks back this quiet error right before the app is finished building:
```
<s> [webpack.Progress] 95% emitting CopyPlugin
<s> [webpack.Progress] 95% emitting inspectpack-duplicates-plugin

Duplicate Sources / Packages - No duplicates found. ðŸš€

<s> [webpack.Progress] 98% after emitting
<s> [webpack.Progress] 98% after emitting CopyPlugin
 ERROR  Failed to compile with 1 errors2:48:50 AM

This dependency was not found:

* habitica-markdown/withMentions in ./src/libs/renderWithMentions.js

To install it, you can run: npm install --save habitica-markdown/withMentions
<s> [webpack.Progress] 100% 
```

So I tried to run `npm install --save habitica-markdown/withMentions` from inside the container. When I did, I see that the container cannot access a habitica git repository.

```
The authenticity of host 'github.com (192.30.255.112)' can't be established.ac72
RSA key fingerprint is SHA256:nThbg6kXUpJWGl7E1IGOCspRomTxdCARLviKw6E5SY8.
[..................] / rollbackFailedOptional: verb npm-session 24fdd25a0357ac72
npm ERR! Error while executing:
npm ERR! /usr/bin/git ls-remote -h -t ssh://git@github.com/habitica-markdown/withMentions.git
npm ERR! 
npm ERR! Host key verification failed.
npm ERR! fatal: Could not read from remote repository.
npm ERR! 
npm ERR! Please make sure you have the correct access rights
npm ERR! and the repository exists.
npm ERR! 
npm ERR! exited with error code: 128

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2020-05-22T03_02_42_848Z-debug.log
The complete log of this npm run:
```

I assume that all the repositories that might be referenced by the build process must be public... right? So, just as a sanity check...

```
$ git ls-remote -h -t ssh://git@github.com/habitica-markdown/withMentions.git
ERROR: Repository not found.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

I'm unsure what is keeping me from downloading that repository... Can you offer any help? 

Thanks!

P.S. 

I'm interested if anyone has a good guide for setting up Webstorm IDE to debug the API server in the docker container on Windows 10? I've had mixed success so far.

--- a/test/api/v3/integration/tasks/PUT-tasks_id.test.js
+++ b/test/api/v3/integration/tasks/PUT-tasks_id.test.js
@@ -499,6 +499,40 @@ describe('PUT /tasks/:id', () => {
     });
   });
 
+  context('monthly dailys', () => {
+    let monthly;
+
+    beforeEach(async () => {
+      const date1 = moment.utc('2020-07-01').toDate()
+      monthly = await user.post('/tasks/user', {
+        text: 'test monthly',
+        type: 'daily',
+        frequency: 'monthly',
+        startDate: date1,
+        daysOfMonth: [date1.getDate()],
+      });
+    });
+
+    it('updates days of month when start date updated', async () => {
+      const date2 = moment.utc('2020-07-01').toDate()
+      const savedMonthly = await user.put(`/tasks/${monthly._id}`, {
+        startDate: date2,
+      });
+
+      expect(savedMonthly.daysOfMonth).to.deep.equal([moment(date2).date()]);
+    });
+
+    it('updates next due when start date updated', async () => {
+      const date2 = moment.utc('2020-07-01').toDate()
+      const savedMonthly = await user.put(`/tasks/${monthly._id}`, {
+        startDate: date2,
+      });
+
+      expect(savedMonthly.nextDue.length).to.eql(6);
+      expect(moment(savedMonthly.nextDue[0]).toDate()).to.eql(moment.utc('2020-08-01').toDate());
+    });
+  });
+
   context('rewards', () => {
     let reward;
 

--- a/test/api/v3/integration/tasks/PUT-tasks_id.test.js
+++ b/test/api/v3/integration/tasks/PUT-tasks_id.test.js
@@ -503,7 +503,7 @@ describe('PUT /tasks/:id', () => {
     let monthly;
 
     beforeEach(async () => {
-      const date1 = moment.utc('2020-07-01').toDate()
+      const date1 = moment.utc('2020-07-01').toDate();
       monthly = await user.post('/tasks/user', {
         text: 'test monthly',
         type: 'daily',
@@ -514,7 +514,7 @@ describe('PUT /tasks/:id', () => {
     });
 
     it('updates days of month when start date updated', async () => {
-      const date2 = moment.utc('2020-07-01').toDate()
+      const date2 = moment.utc('2020-07-01').toDate();
       const savedMonthly = await user.put(`/tasks/${monthly._id}`, {
         startDate: date2,
       });
@@ -523,13 +523,18 @@ describe('PUT /tasks/:id', () => {
     });
 
     it('updates next due when start date updated', async () => {
-      const date2 = moment.utc('2020-07-01').toDate()
+      const date2 = moment.utc('2020-07-01').toDate();
       const savedMonthly = await user.put(`/tasks/${monthly._id}`, {
         startDate: date2,
       });
 
       expect(savedMonthly.nextDue.length).to.eql(6);
       expect(moment(savedMonthly.nextDue[0]).toDate()).to.eql(moment.utc('2020-08-01').toDate());
+      expect(moment(savedMonthly.nextDue[1]).toDate()).to.eql(moment.utc('2020-09-01').toDate());
+      expect(moment(savedMonthly.nextDue[2]).toDate()).to.eql(moment.utc('2020-10-01').toDate());
+      expect(moment(savedMonthly.nextDue[3]).toDate()).to.eql(moment.utc('2020-11-01').toDate());
+      expect(moment(savedMonthly.nextDue[4]).toDate()).to.eql(moment.utc('2020-12-01').toDate());
+      expect(moment(savedMonthly.nextDue[5]).toDate()).to.eql(moment.utc('2021-01-01').toDate());
     });
   });
 

--- a/website/server/controllers/api-v3/tasks.js
+++ b/website/server/controllers/api-v3/tasks.js
@@ -678,6 +678,10 @@ api.updateTask = {
       task.group.sharedCompletion = sanitizedObj.sharedCompletion;
     }
 
+    if (task.frequency === 'monthly' && task) {
+      task.daysOfMonth = [ moment(task.startDate).date() ];
+    }
+
     setNextDue(task, user);
     const savedTask = await task.save();
 

--- a/website/server/controllers/api-v3/tasks.js
+++ b/website/server/controllers/api-v3/tasks.js
@@ -678,7 +678,12 @@ api.updateTask = {
       task.group.sharedCompletion = sanitizedObj.sharedCompletion;
     }
 
-    if (task.frequency === 'monthly' && task.startDate) {
+    // If the task was set to repeat monthly on a day of the month, and the start date was updated,
+    // the task will then need to be updated to repeat on the same day of the month as the new
+    // start date.  For example, if the start date is updated to 7/2/2020, the daily should
+    // repeat on the 2nd day of the month.  It's possible that a task can repeat monthly on a 
+    // week of the month, in which case we won't update the repetition at all.
+    if (task.frequency === 'monthly' && task.daysOfMonth.length && task.startDate) {
       task.daysOfMonth = [moment(task.startDate).date()];
     }
 

--- a/website/server/controllers/api-v3/tasks.js
+++ b/website/server/controllers/api-v3/tasks.js
@@ -678,8 +678,8 @@ api.updateTask = {
       task.group.sharedCompletion = sanitizedObj.sharedCompletion;
     }
 
-    if (task.frequency === 'monthly' && task) {
-      task.daysOfMonth = [ moment(task.startDate).date() ];
+    if (task.frequency === 'monthly' && task.startDate) {
+      task.daysOfMonth = [moment(task.startDate).date()];
     }
 
     setNextDue(task, user);

--- a/website/server/controllers/api-v3/tasks.js
+++ b/website/server/controllers/api-v3/tasks.js
@@ -680,8 +680,8 @@ api.updateTask = {
 
     // If the task was set to repeat monthly on a day of the month, and the start date was updated,
     // the task will then need to be updated to repeat on the same day of the month as the new
-    // start date.  For example, if the start date is updated to 7/2/2020, the daily should
-    // repeat on the 2nd day of the month.  It's possible that a task can repeat monthly on a 
+    // start date. For example, if the start date is updated to 7/2/2020, the daily should
+    // repeat on the 2nd day of the month. It's possible that a task can repeat monthly on a
     // week of the month, in which case we won't update the repetition at all.
     if (task.frequency === 'monthly' && task.daysOfMonth.length && task.startDate) {
       task.daysOfMonth = [moment(task.startDate).date()];

